File: dsa_leetcode_beginner/0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:   
        
        # Initialize the boundaries
        rowBegin, rowEnd = 0, len(matrix) - 1
        colBegin, colEnd = 0, len(matrix[0]) - 1
        
        # Resultant list
        result = []
        
        while rowBegin <= rowEnd and colBegin <= colEnd:
            # colBegin to colEnd along the rowBegin boundary
            for i in range(colBegin, colEnd + 1):
                result.append(matrix[rowBegin][i])
            rowBegin += 1
            
            # rowBegin to rowEnd along the colEnd boundary
            for i in range(rowBegin, rowEnd + 1):
                result.append(matrix[i][colEnd])
            colEnd -= 1
            
            if rowBegin <= rowEnd:
                # colEnd to colBegin along the rowEnd boundary
                for i in range(colEnd, colBegin - 1, -1):
                    result.append(matrix[rowEnd][i])
                rowEnd -= 1
            
            if colBegin <= colEnd:
                # rowEnd to rowBegin along the colBegin boundary
                for i in range(rowEnd, rowBegin - 1, -1):
                    result.append(matrix[i][colBegin])
                colBegin += 1
        
        return result
    

#Question: https://leetcode.com/problems/spiral-matrix
#Blog: https://blog.unwiredlearning.com/spiral-matrix

File: dsa_leetcode_beginner/0054-spiral-matrix/0054-spiral-matrix.js

function spiralOrder(matrix) {
    let result = [];
    if (!matrix || matrix.length === 0) return result;

    let rowBegin = 0, rowEnd = matrix.length - 1;
    let colBegin = 0, colEnd = matrix[0].length - 1;

    while (rowBegin <= rowEnd && colBegin <= colEnd) {
        for (let i = colBegin; i <= colEnd; i++) {
            result.push(matrix[rowBegin][i]);
        }
        rowBegin++;

        for (let i = rowBegin; i <= rowEnd; i++) {
            result.push(matrix[i][colEnd]);
        }
        colEnd--;

        if (rowBegin <= rowEnd) {
            for (let i = colEnd; i >= colBegin; i--) {
                result.push(matrix[rowEnd][i]);
            }
            rowEnd--;
        }

        if (colBegin <= colEnd) {
            for (let i = rowEnd; i >= rowBegin; i--) {
                result.push(matrix[i][colBegin]);
            }
            colBegin++;
        }
    }

    return result;
}

// Example usage:
let matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]];
console.log(spiralOrder(matrix));
